python/cudf/cudf/core/frame.py

```diff
@@ -3337,18 +3329,240 @@ def _reindex(
 
         return self._mimic_inplace(result, inplace=inplace)
 
+    def _unaryop(self, op):
+        data_columns = (col.unary_operator(op) for col in self._columns)
+        data = zip(self._column_names, data_columns)
+        return self.__class__._from_table(Frame(data, self._index))
+
+    def _binaryop(
+        self,
+        other: T,
+        fn: str,
+        fill_value: Any = None,
+        reflect: bool = False,
+        *args,
+        **kwargs,
+    ) -> Frame:
+        raise NotImplementedError
+
+    @classmethod
+    def _colwise_binop(
+        cls,
+        operands: Dict[Optional[str], Tuple[ColumnBase, Any, bool, Any]],
+        fn: str,
+    ):
+        """Implement binary ops between two frame-like objects.
+
+        Binary operations for Frames can be reduced to a sequence of binary
+        operations between column-like objects. Different types of frames need
+        to preprocess different inputs, so subclasses should implement binary
+        operations as a preprocessing step that calls this method.
+
+        Parameters
+        ----------
+        operands : Dict[Optional[str], Tuple[ColumnBase, Any, bool, Any]]
+            A mapping from column names to a tuple containing left and right
+            operands as well as a boolean indicating whether or not to reflect
+            an operation and fill value for nulls.
+        fn : str
+            The operation to perform.
+
+        Returns
+        -------
+        Frame
+            A subclass of Frame constructed from the result of performing the
+            requested operation on the operands.
+        """
+
+        # Now actually perform the binop on the columns in left and right.
+        output = {}
+        for (
+            col,
+            (left_column, right_column, reflect, fill_value),
+        ) in operands.items():
+            if right_column is cudf.NA:
```

Comment:
might consider keeping a dict of these null scalars around keyed by dtype for the lifetime of this function. Might help perf in the wide cases.

Reply:
Isn't this object already an instance rather than something that has to be created each time? It looks like it's created only once: https://github.com/rapidsai/cudf/blob/branch-21.08/python/cudf/cudf/core/scalar.py#L393

Reply:
Oh sorry, comment is a line too high. This refers to the `cudf.Scalar` objects that we make next. It's certainly just an optimization, but if you already happen to have created and used a `cudf.Scalar` with value `<NA>` that has the `dtype` you need for a subsequent op, you can just reuse that and it won't need to allocate and construct it again.

It could even make sense for us to cache a null device scalar for each of our scalar types as they get used, but that's definitely out of scope.

Reply:
Discussed offline, we're going to punt on this for now.